### DISCUSSION TITLE
Linter for common gotchas

### DISCUSSION THREAD

**Discussion by rassie** [sqlalchemy/sqlalchemy#9050] [Ideas]
I'm currently in the process of migrating my SQLAlchemy 1.3 project to async DB driver and thus updating to SQLAlchemy 2.0, since I must use the new 2.0-style query syntax and while I'm at it, I'd like to use typings more. 

Now here is the thing: it's quite a lot of stuff to keep in mind. SA's documentation is excellent, but it's also vast. I find myself constantly trying to remember to move `Column`s to `mapped_column`s, to add `Mapped` everywhere where it's appropriate, to add `limit(1)` on `.first()` queries and so on. There are a thousand other things that SA provides and a lot of gotchas and inefficiencies that more or less require a seasoned developer to tackle.

Is there any chance we could see a linter or a linter plugin ( (`flake8` or `ruff` come to mind) for SQLAlchemy stuff, which would point newbies and veterans alike to possible errors they've made or improvements they could be making?

**Comment by zzzeek** [comment]
hi -

why not use dmypy directly?   if your models pass type checking fully, then you've used all the new APIs completely.

**Reply by rassie** [reply]
Already doing that, it's a big help, that doesn't fully help with more optional things you have to think about e.g. using mapped columns, typing them correctly and fully, using integrated UUID type instead of one from sqlalchemy_utils etc.

**Reply by zzzeek** [reply]
what kind of linter would we provide that looks for SQLAlchemy's Uuid datatype and is hardcoded to both know about sqlalchemy_utils as well as to discourage using it?   I'm not really sure how a linter would be expected to know that kind of thing.

**Reply by zzzeek** [reply]
also re: using mapped_column, if you are using Declarative, your code will not pass mypy typing if you are putting Column in your declarative models and not mapped_column.   that should already be a solved problem   (edit: and you are also using those models in queries)

**Reply by zzzeek** [reply]
obviously disable the legacy SQLAlchemy mypy plugin

**Reply by zzzeek** [reply]
re: Column the only way I could think of catching that is to just disallow its use in Declarative, that is, a runtime check.  Likely just change your imports to remove "from sqlalchemy import Column".

basically I dont see us writing a source code linter on this end, that would be out of scope for what we can provide here.   Certainly people can create and publish these, however.
